Adding policy for CPU Limits for all container types

nsagark opened the pull request:
## Related Issue(s)

<!--
Please link the GitHub issue this pull request resolves by using the appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
-->

## Description

<!--
Adding policy for CPU Limits for all container types
-->

## Checklist

<!--
Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for.
-->

- [x] I have read the [policy contribution guidelines](https://github.com/kyverno/policies/blob/main/README.md#contribution).
- [x] I have added test manifests and resources covering both positive and negative tests that prove this policy works as intended.
- [x] I have added the artifacthub-pkg.yml file and have verified it is complete and correct.

chipzoller commented on other/require-cpu-limits/require-cpu-limits.yaml:
```diff
@@ -0,0 +1,40 @@
+  background: true
+  rules:
+  - name: check-cpu-limits
+    match:
+      resources:
```
All policies should be using `any`/`all` under the match and exclude conditions.

chipzoller commented on other/require-cpu-limits/require-cpu-limits.yaml:
```diff
@@ -0,0 +1,40 @@
+    policies.kyverno.io/subject: Pod
+    kyverno.io/kubernetes-version: "1.26"
+    policies.kyverno.io/description: >-
+      Setting CPU limits on pods ensures fair distribution of resources, preventing any single pod from monopolizing CPU and impacting the performance of other pods. This practice enhances stability, predictability, and cost control, while also mitigating the noisy neighbor problem and facilitating efficient scaling in Kubernetes clusters. This policy ensures that cpu limits are set on every cont ...
```
CPU limits aren't set on Pods, they're set on containers.

chipzoller commented on other/require-cpu-limits/.chainsaw-test/pod-bad.yaml:
This test resource manifest needs to cover cases where there are multiple containers (both `containers[]` and `initContainers[]`) in which one container sets a limit while the other does not. You also do not need to include any securityContext fields as these are irrelevant to this policy.

chipzoller commented on other/require-cpu-limits/.kyverno-test/resource.yaml:
See similar comment on Chainsaw resource file.

nsagark commented:
@chipzoller I have updated the files as per your comments. Please review

chipzoller commented:
Are all comments resolved?

nsagark commented:
Hi @chipzoller Yes, I have updated the files. Just to make sure to address below comment,

`This test resource manifest needs to cover cases where there are multiple containers (both containers[] and initContainers[]) in which one container sets a limit while the other does not`

I have added below in resource.yaml file. Is that what you meant or something else?

`apiVersion: v1
kind: Pod
metadata:
  name: badpod07
spec:
  initContainers:
  - name: initcontainer01
    image: dummyimagename
  - name: initcontainer02
    image: dummyimagename
  containers:
  - name: container01
    image: dummyimagename
    resources:
      limits:
        cpu: "500m"
  - name: container02
    image: dummyimagename    `

chipzoller commented:
Yes

nsagark commented:
With that, I have addressed all the comments. Please review.

chipzoller commented on other/require-cpu-limits/.chainsaw-test/pod-bad.yaml:
```diff
@@ -0,0 +1,84 @@
+  initContainers:
+  - name: initcontainer01
+    image: busybox:1.35
+  - name: initcontainer02
+    image: busybox:1.35
```
Please update this initcontainer with a CPU limit.

chipzoller commented on other/require-cpu-limits/.chainsaw-test/pod-bad.yaml:
```diff
@@ -0,0 +1,84 @@
+  - name: container01
+    image: busybox:1.35
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: badpod02
+spec:
+  containers:
+  - name: container01
+    image: busybox:1.35
```
This is redundant from badpod01. Remove.

chipzoller commented on other/require-cpu-limits/.chainsaw-test/pod-bad.yaml:
```diff
@@ -0,0 +1,84 @@
+  - name: container02
+    image: busybox:1.35
+---
+apiVersion: v1
+kind: Pod
+metadata:
+  name: badpod04
+spec:
+  containers:
+  - name: container01
+    image: busybox:1.35
+  - name: container02
+    image: busybox:1.35
```
This is redundant from badpod03. Remove.

chipzoller commented on other/require-cpu-limits/.chainsaw-test/podcontroller-bad.yaml:
Please make similar changes to this file as those pointed out in `pod-bad.yaml`.

chipzoller reviewed (changes requested):
Please ensure to resolve all conversations when complete.

nsagark commented:
@chipzoller All conversations are resolved.

chipzoller commented:
Several conversations unresolved.

nsagark commented:
@chipzoller Sorry, I have resolved previous conversations. Please check now

chipzoller reviewed (changes requested):
Looks good other than the AH digest needing to be updated.

nsagark commented:
Hi @chipzoller The AH digest seems to be correct. Am I doing anything wrong below?

Sagar@DESKTOP-VS85098 MINGW64 /c/users/sagar/Downloads/kyverno-prs/policies/other/require-cpu-limits (require-cpu-limits)
$ sha256sum require-cpu-limits.yaml
2fe9fa043fbb4c85da0204a00a60ce6732ac0dc0e3a5ece4f7b741982a1003d1 *require-cpu-limits.yaml

Sagar@DESKTOP-VS85098 MINGW64 /c/users/sagar/Downloads/kyverno-prs/policies/other/require-cpu-limits (require-cpu-limits)
$ ls -lrht
total 8.0K
-rw-r--r-- 1 Sagar 197121 1.5K Jul 31 09:47 require-cpu-limits.yaml
-rw-r--r-- 1 Sagar 197121 1.5K Jul 31 09:47 artifacthub-pkg.yml

chipzoller commented:
You're not doing anything wrong, it's just the expected digest is calculated in the pipeline under Linux. You're doing it under Windows which means the digest may be slightly different. The one being reported in the pipeline is `1d9998010342080ae64f309befdf52065de557cdcc10ddf68d8476b5af93d505`. Go ahead and put that in the AH metadata file and it should pass.

nsagark commented:
Hi @chipzoller I have updated the digest. Thanks for pointing it out.

nsagark commented:
Hi @chipzoller I don't see the changes yet in the main branch. Are we good here or anything missing?
